internal/core/service/user_service.go

```diff
@@ -0,0 +1,148 @@
+package service
+
+import (
+	"errors"
+	"github.com/gin-gonic/gin"
+	"github.com/viniciusgferreira/ps-tag-onboarding-go/internal/core/domain/model"
+	"log/slog"
+	"regexp"
+)
+
+var (
+	ErrUserNotFound      = errors.New("user not found")
+	ErrUserAlreadyExists = errors.New("user with the same first and last name already exists")
+	ErrInvalidAge        = errors.New("user must be at least 18 years old")
+	ErrInvalidName       = errors.New("user first and last name cannot be empty")
+	ErrInvalidEmail      = errors.New("invalid email")
+)
+
+type UserRepository interface {
+	FindById(ctx *gin.Context, id string) (*model.User, error)
+	Save(ctx *gin.Context, u model.User) (*model.User, error)
+	Update(ctx *gin.Context, u model.User) (*model.User, error)
+	ExistsByFirstNameAndLastName(ctx *gin.Context, u model.User) (bool, error)
+}
+type UserService struct {
+	repo UserRepository
+}
+
+func NewUserService(repo UserRepository) *UserService {
+	return &UserService{repo: repo}
+}
+
+type ValidationError struct {
+	Message string
+	Details []string
+}
+
+func NewValidationErrorWithDetails(validationErrors []error) ValidationError {
+	details := make([]string, len(validationErrors))
+	for i, err := range validationErrors {
+		details[i] = err.Error()
+	}
+	return ValidationError{
+		Message: "user did not pass validation",
+		Details: details,
+	}
+}
+
+func (r ValidationError) Error() string {
+	return r.Message
+}
+
+func (s *UserService) Find(ctx *gin.Context, id string) (*model.User, error) {
```

Comment:
All of your functions in this file and repository can receive a context.Context instead of a gin.Context.
That's because we want to decouple all layers one from another, if you do this way and change your handler framework, you will need to change everything else.